File: myExpenses/src/main/java/org/totschnig/myexpenses/model/PreferencesCurrencyContext.java
package org.totschnig.myexpenses.model;

import android.os.Build;

import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.util.locale.UserLocaleProvider;

import java.util.Collections;
import java.util.Currency;
import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;

public class PreferencesCurrencyContext implements CurrencyContext {
  /**
   * used with currencies where Currency.getDefaultFractionDigits returns -1
   */
  public static final int DEFAULTFRACTIONDIGITS = 8;
  private static final String KEY_CUSTOM_FRACTION_DIGITS = "CustomFractionDigits";
  private static final String KEY_CUSTOM_CURRENCY_SYMBOL = "CustomCurrencySymbol";
  final private PrefHandler prefHandler;
  final private UserLocaleProvider userLocaleProvider;
  private static final Map<String, CurrencyUnit> INSTANCES = Collections.synchronizedMap(new HashMap<>());

  public PreferencesCurrencyContext(PrefHandler prefHandler, UserLocaleProvider userLocaleProvider) {
    this.prefHandler = prefHandler;
    this.userLocaleProvider = userLocaleProvider;
  }

  @Override
  @NonNull
  public CurrencyUnit get(@NonNull String currencyCode) {
    synchronized (this) {
      CurrencyUnit currencyUnit = INSTANCES.get(currencyCode);
      if (currencyUnit != null) {
        return currencyUnit;
      }

      Currency c = Utils.getInstance(currencyCode);
      if (c != null) {
        currencyUnit = new CurrencyUnit(currencyCode, getSymbol(c), getFractionDigits(c), getDisplayName(c));
      } else {
        final String customSymbol = getCustomSymbol(currencyCode);
        final int customFractionDigits = getCustomFractionDigits(currencyCode);
        currencyUnit = new CurrencyUnit(currencyCode, customSymbol == null ? "¤" : customSymbol,
            customFractionDigits == -1 ? DEFAULTFRACTIONDIGITS : customFractionDigits);
      }
      INSTANCES.put(currencyCode, currencyUnit);
      return currencyUnit;
    }
  }

  private String getDisplayName(Currency c) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
      return c.getDisplayName();
    } else {
      return c.getCurrencyCode();
    }
  }

  public String getCustomSymbol(String currencyCode) {
    return prefHandler.getString(currencyCode + KEY_CUSTOM_CURRENCY_SYMBOL, null);
  }

  public int getCustomFractionDigits(String currencyCode) {
    return prefHandler.getInt(currencyCode + KEY_CUSTOM_FRACTION_DIGITS, -1);
  }

  public String getSymbol(@NonNull Currency currency) {
    String custom = getCustomSymbol(currency.getCurrencyCode());
    return custom != null ? custom : currency.getSymbol(userLocaleProvider.getUserPreferredLocale());
  }

  public int getFractionDigits(Currency currency) {
    int customFractionDigits = getCustomFractionDigits(currency.getCurrencyCode());
    if (customFractionDigits != -1) {
      return customFractionDigits;
    }
    int digits = currency.getDefaultFractionDigits();
    if (digits != -1) {
      return digits;
    }
    return DEFAULTFRACTIONDIGITS;
  }

  @Override
  public void storeCustomFractionDigits(String currencyCode, int fractionDigits) {
    prefHandler.putInt(currencyCode + KEY_CUSTOM_FRACTION_DIGITS, fractionDigits);
    INSTANCES.remove(currencyCode);
  }

  @Override
  public void storeCustomSymbol(String currencyCode, String symbol) {
    Currency currency = null;
    try {
      currency = Currency.getInstance(currencyCode);
    } catch (Exception ignored) {
    }
    String key = currencyCode + KEY_CUSTOM_CURRENCY_SYMBOL;
    if (currency != null && currency.getSymbol().equals(symbol)) {
      prefHandler.remove(key);
    } else {
      prefHandler.putString(key, symbol);
    }
    INSTANCES.remove(currencyCode);
  }

  @Override
  public void ensureFractionDigitsAreCached(CurrencyUnit currency) {
    storeCustomFractionDigits(currency.getCode(), currency.getFractionDigits());
  }

  @Override
  public void invalidateHomeCurrency() {
    INSTANCES.remove(AggregateAccount.AGGREGATE_HOME_CURRENCY_CODE);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/model/PreferencesCurrencyContext.kt
package org.totschnig.myexpenses.model

import androidx.annotation.VisibleForTesting
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.provider.DataBaseAccount
import org.totschnig.myexpenses.util.Utils
import java.util.Collections
import java.util.Currency
import java.util.Locale

open class PreferencesCurrencyContext(
    private val prefHandler: PrefHandler,
    private val application: MyApplication
) : CurrencyContext {
    override fun get(currencyCode: String): CurrencyUnit {
        if (currencyCode == DataBaseAccount.AGGREGATE_HOME_CURRENCY_CODE) return homeCurrencyUnit
        synchronized(this) {
            var currencyUnit = INSTANCES[currencyCode]
            if (currencyUnit != null) {
                return currencyUnit
            }
            val c = Utils.getInstance(currencyCode)
            currencyUnit = if (c != null) {
                CurrencyUnit(currencyCode, getSymbol(c), getFractionDigits(c), c.displayName)
            } else {
                val customSymbol = getCustomSymbol(currencyCode)
                val customFractionDigits = getCustomFractionDigits(currencyCode)
                CurrencyUnit(
                    currencyCode, customSymbol ?: "¤",
                    if (customFractionDigits == -1) DEFAULT_FRACTION_DIGITS else customFractionDigits
                )
            }
            INSTANCES[currencyCode] = currencyUnit
            return currencyUnit
        }
    }

    private fun getCustomSymbol(currencyCode: String) =
        prefHandler.getString(customSymbolKey(currencyCode), null)

    private fun getCustomFractionDigits(currencyCode: String) =
        prefHandler.getInt(customFractionDigitsKey(currencyCode), -1)

    private fun getSymbol(currency: Currency) = getCustomSymbol(currency.currencyCode)
        ?: currency.getSymbol(application.userPreferredLocale)

    private fun getFractionDigits(currency: Currency) =
        getCustomFractionDigits(currency.currencyCode).takeIf { it != -1 } ?:
        currency.getDefaultFractionDigits().takeIf { it != -1 } ?: DEFAULT_FRACTION_DIGITS

    override fun storeCustomFractionDigits(currencyCode: String, fractionDigits: Int) {
        prefHandler.putInt(customFractionDigitsKey(currencyCode), fractionDigits)
        INSTANCES.remove(currencyCode)
    }

    override fun storeCustomSymbol(currencyCode: String, symbol: String) {
        val currency = try {
            Currency.getInstance(currencyCode)
        } catch (_: Exception) { null }
        val key = customSymbolKey(currencyCode)
        if (currency != null && currency.symbol == symbol) {
            prefHandler.remove(key)
        } else {
            prefHandler.putString(key, symbol)
        }
        INSTANCES.remove(currencyCode)
    }

    override fun ensureFractionDigitsAreCached(currency: CurrencyUnit) {
        storeCustomFractionDigits(currency.code, currency.fractionDigits)
    }

    override fun invalidateHomeCurrency() {
        INSTANCES.remove(DataBaseAccount.AGGREGATE_HOME_CURRENCY_CODE)
    }

    override val homeCurrencyString: String
        get() = prefHandler.getString(PrefKey.HOME_CURRENCY, null) ?: localCurrency.currencyCode

    override val localCurrency = Utils.getCountryFromTelephonyManager(application)?.let {
        try {
            Currency.getInstance(Locale("", it))
        } catch (_: Exception) {
            null
        }
    } ?: Utils.getSaveDefault()

    companion object {
        /**
         * used with currencies where Currency.getDefaultFractionDigits returns -1
         */
        const val DEFAULT_FRACTION_DIGITS = 8
        private const val KEY_CUSTOM_FRACTION_DIGITS = "CustomFractionDigits"
        private const val KEY_CUSTOM_CURRENCY_SYMBOL = "CustomCurrencySymbol"
        private val INSTANCES = Collections.synchronizedMap(HashMap<String?, CurrencyUnit>())

        private fun customFractionDigitsKey(currencyCode: String) = currencyCode + KEY_CUSTOM_FRACTION_DIGITS
        private fun customSymbolKey(currencyCode: String) = currencyCode + KEY_CUSTOM_CURRENCY_SYMBOL

        @VisibleForTesting
        fun resetFractionDigits(prefHandler: PrefHandler, currencyCode: String) {
            prefHandler.remove(customFractionDigitsKey(currencyCode))
            INSTANCES.remove(currencyCode)
        }
    }
}
